Add DOCUMENTATION.md (#667)
rapidsai/cuspatial: cpp/doc/DOCUMENTATION.md

  `2019-2021`)

Changing the copyright year may not be necessary if no content has changed (e.g. reformatting only).

## Doxygen

The [doxygen tool](http://www.doxygen.nl/manual/index.html) is used to generate HTML pages from the
C++ comments in the source code. Doxygen recognizes and parses block comments and performs
specialized output formatting when it encounters
[doxygen commands](http://www.doxygen.nl/manual/commands.html).

There are almost 200 commands (also called tags in this document) that doxygen recognizes in comment
blocks. This document provides guidance on which commands/tags to use and how to use them in the
libcuspatial C++ source code.

The doxygen process can be customized using options in the [Doxyfile](../doxygen/Doxyfile).

Here are some of the custom options in the Doxyfile for libcuspatial.
| Option | Setting | Description |
| ------ | ------- | ----------- |
| PROJECT_NAME | libcuspatial | Title used on the main page |
| PROJECT_NUMBER | 22.10.00 | Version number |
| EXTENSION_MAPPING | cu=C++ cuh=C++ | Process `cu` and `cuh` as C++ |
| INPUT | main_page.md ../include | Embedded markdown files and source code directories to process |

> | FILE_PATTERNS | *.cpp *.hpp *.h *.c *.cu *.cuh | File extensions to process |

Author: > And `TAGFILES` where we configure external links to rmm and libcudf. I think it gets automatically updated per release?

Oh, it wasn't. But I added code to update it in #673 (just seemed to fit that PR better).

> | FILE_PATTERNS | *.cpp *.hpp *.h *.c *.cu *.cuh | File extensions to process |

Author: Added all these.

code in libcuspatial.

```c++
/**
 * @file source_file.cpp
 * @brief Description of source file contents
 *
 * Longer description of the source file contents.
 */

/**
 * @brief One sentence description of the class.
 *
 * @ingroup optional_predefined_group_id
 *
 * Longer, more detailed description of the class.
 *
 * @tparam T Short description of each template parameter
 * @tparam U Short description of each template parameter
 */
template <typename T, typename U>
class example_class {

  void get_my_int();            ///< Simple members can be documented like this
  void set_my_int( int value ); ///< Try to use descriptive member names

  /**
   * @brief Short, one sentence description of the member function.
   *
   * A more detailed description of what this function does and what
   * its logic does.
   *
   * @code
   * example_class<int> inst;
   * inst.set_my_int(5);
   * int output = inst.complicated_function(1,dptr,fptr);
   * @endcode
   *
   * @param[in]     first  This parameter is an input parameter to the function
   * @param[in,out] second This parameter is used both as an input and output

>    * @param[out]    third  This parameter is an output of the function

Author: Added discussion of this.

parameter passed to this function. The name of the parameter specified after the doxygen tag must
match the function's parameter name. Also include append `[in]`, `[out]` or `[in,out]` to the
`@param` if it is not clear from the declaration and the parameter name itself.

```c++
 /**
 * ...
 * @param[in]     first  This parameter is an input parameter to the function
 * @param[in,out] second This parameter is used both as an input and output
 * @param[out]    third  This parameter is an output of the function
 */
```

It is also recommended to vertically align the 3 columns of text if possible to make it easier to
read in a source code editor.

#### @return

Add a single [@return](http://www.doxygen.nl/manual/commands.html#cmdreturn) comment line at the end
of the comment block if the function returns an object or value. Include a brief description of what
is returned.

```c++
/**
 * ...
 *
 * @return A new column of type INT32 and no nulls
 */
```

Do not include the type of the object returned with the `@return` comment.

#### @pre

Add a [@pre](https://www.doxygen.nl/manual/commands.html#cmdpre) comment line for each precondition

> of the function. For example, assumptions about accessibility or range of iterators.

Author: Well for example device accessibility. We use @pre to state that iterator ranges may or may not overlap (just as Thrust does), which is definitely a run-time requirement.

> of the function. For example, assumptions about accessibility or range of iterators.

Author: Added discussion of this.

`@addtogroup` with `@{ ... @}` pair includes a namespace declaration. So include the `@addtogroup`
and `@{ ... @}` between the namespace declaration braces as shown in the example above.

Summary of groups tags
| Tag/Command | Where to use |
| ----------- | ------------ |
| `@defgroup` | For use only in [doxygen_groups.h](../include/doxygen_groups.h) and should include the group's title. |
| `@ingroup` | Use inside individual doxygen block comments for declaration statements in a header file. |
| `@addtogroup` | Use instead of `@ingroup` for multiple declarations in the same file within a namespace declaration. Do not specify a group title. |
| `@{ ... @}` |  Use only with `@addtogroup`. |

## Build Doxygen Output

We recommend installing Doxygen using conda (`conda install doxygen`) or a Linux package manager
(`sudo apt install doxygen`). Alternatively you can
[build and install doxygen from source](http://www.doxygen.nl/manual/install.html).

To build the libcuspatial HTML documentation simply run the `doxygen` command from the `cpp/doxygen`
directory containing the `Doxyfile`. The libcuspatial documentation can also be built using

> `make docs_cuspatial` from the cmake build directory (e.g. `cpp/build`). Doxygen reads and processes

Reviewer: I actually have issue executing this command. I think it's also what leads to `docs-cuspatial-cpp` not working in my compose. @trxcllnt

Reviewer: ```
make docs_cuspaital
make: *** No rule to make target 'docs_cuspaital'.  Stop.
```

Author: You have a typo. Did you try `make docs_cuspatial`?

Author: Removing this. Doesn't seem to exist.

Author: Re-added and clarified, and added the missing targets to the CMake. Now need a cmake approval.

 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
```

The comment should start with `/*` and not `/**` so it is not processed by doxygen.

Also, here are the rules for the copyright year.

- A new file should have the year in which it was created
- A modified file should span the year it was created and the year it was modified (e.g.
  `2019-2021`)

Changing the copyright year may not be necessary if no content has changed (e.g. reformatting only).

## Doxygen

The [doxygen tool](http://www.doxygen.nl/manual/index.html) is used to generate HTML pages from the
C++ comments in the source code. Doxygen recognizes and parses block comments and performs
specialized output formatting when it encounters
[doxygen commands](http://www.doxygen.nl/manual/commands.html).

There are almost 200 commands (also called tags in this document) that doxygen recognizes in comment

> blocks. This document provides guidance on which commands/tags to use and how to use them in the

Reviewer: ```suggestion
blocks. This [descriptions](#descriptions) section provides guidance on which commands/tags to use and how to use them in the
```

Author: I don't see the section you are linking to. I think referring to "this document" is fine.

libcuspatial C++ source code.

> The doxygen process can be customized using options in the [Doxyfile](../doxygen/Doxyfile).

Reviewer: ```suggestion
Doxygen can be customized using options in the [Doxyfile](../doxygen/Doxyfile).
```

 *
 * Longer, more detailed description of the class.
 *
 * @tparam T Short description of each template parameter
 * @tparam U Short description of each template parameter
 */
template <typename T, typename U>
class example_class {

  void get_my_int();            ///< Simple members can be documented like this
  void set_my_int( int value ); ///< Try to use descriptive member names

  /**
   * @brief Short, one sentence description of the member function.
   *
   * A more detailed description of what this function does and what
   * its logic does.
   *
   * @code
   * example_class<int> inst;
   * inst.set_my_int(5);
   * int output = inst.complicated_function(1,dptr,fptr);
   * @endcode
   *
   * @param[in]     first  This parameter is an input parameter to the function
   * @param[in,out] second This parameter is used both as an input and output
   * @param[out]    third  This parameter is an output of the function
   *
   * @return The result of the complex function
   */
  T complicated_function(int first, double* second, float* third)
  {
      // Do not use doxygen-style block comments
      // for code logic documentation.
  }

 private:

>   int my_int;                ///< An example private member variable

Reviewer: Would be nice to provide an example for "between the line" doxygen comments, like
```suggestion
  int my_int;                ///< An example private member variable
  /// Another example of private member variable
  float my_float;
```
Not that single line comment will automatically be treated as @brief and mutliple line comments are viewed as detail.
